token-tracker/src/truth-sources/token-staking.js

```diff
@@ -0,0 +1,187 @@
+      this.context.web3,
+      this.tokenStaking,
+      "StakeDelegated",
+      this.context.deploymentBlock,
+      this.finalBlock
+    )
+    logger.info(`found ${events.length} stake delegated events`)
+
+    const operatorsOwnersMap = new Map()
+    events.forEach((event) => {
+      operatorsOwnersMap.set(
+        event.returnValues.operator,
+        event.returnValues.owner
+      )
+    })
+
+    dumpDataToFile(operatorsOwnersMap, TOKEN_STAKING_HISTORIC_STAKERS_DUMP_PATH)
+
+    return operatorsOwnersMap
+  }
+
+  /**
+   * Filters owners based on the rules defined for sources of truth. Ignores
+   * addresses of known Keep contracts for which actual holders are resolved.
+   * @param {Map<Address,Address>} operatorsOwnersMap Map of operators and their
+   * owners to filter.
+   * @return {Map<Address,Address>} Filtered map of operators and owners.
+   */
+  async filterOwners(operatorsOwnersMap) {
+    logger.info(`filter owners`)
+
+    const filteredOperatorsOwners = new Map()
+    const unknownContracts = new Set()
+
+    for (const [operator, owner] of operatorsOwnersMap) {
+      const [
+        isIgnored,
+        addressType,
+      ] = await this.addressesResolver.isIgnoredAddress(owner)
+
+      logger.debug(
+        `operator's [${operator}] owner [${owner}] is ${addressType}`
+      )
+
+      if (isIgnored) {
+        continue // skip this operator
+      }
+
+      // If the address doesn't match any of ignored contracts store it for
+      // reference if we want to double check them.
+      if (addressType === AddressesResolver.UNKNOWN_CONTRACT)
+        unknownContracts.add(owner)
+
+      filteredOperatorsOwners.set(operator, owner)
+    }
+
+    dumpDataToFile(
+      unknownContracts,
+      TOKEN_STAKING_UNKNOWN_OWNERS_CONTRACTS_DUMP_PATH
+    )
+
+    return filteredOperatorsOwners
+  }
+
+  /**
+   * Checks token balances for owners in the operators to owner map. It walks over
+   * delegations to operators, ignored delegations that were undelegated. Combines
+   * results for owners that have multiple operators.
+   * @param {Map<Address,Address>} stakers Map of operators and owners to check.
+   * @return {Map<Address,BN>} Token holdings at the final blocks.
+   */
+  async checkStakedValues(stakers) {
+    logger.info(`check stake delegations at block ${this.finalBlock}`)
+
+    /** @type {Map<Address,BN>} */
+    const stakersBalances = new Map()
+
+    for (const [operator, owner] of stakers) {
+      const delegationInfo = await callWithRetry(
+        this.tokenStaking.methods.getDelegationInfo(operator),
+        undefined,
+        undefined,
+        this.finalBlock
+      )
+
+      const amount = toBN(delegationInfo.amount)
+      const undelegatedAt = toBN(delegationInfo.undelegatedAt)
+
+      if (undelegatedAt.gtn(0)) {
+        logger.debug(
+          `skipping delegation to ${operator}, undelegated at: ${undelegatedAt.toString()}`
+        )
+        continue
+      }
```

Comment:
Should we skip them?

`undelegatedAt` is set when the delegation owner starts undelegation:
https://github.com/keep-network/keep-core/blob/c9941e7923f03343eec1fc7f12e2d1d56bbbe262/solidity/contracts/TokenStaking.sol#L249-L251

Still, tokens are still staked at this point. When stake is recovered from `TokenStaking` contract, the amount is set to 0 but `undelegatedAt` is not updated:

https://github.com/keep-network/keep-core/blob/c9941e7923f03343eec1fc7f12e2d1d56bbbe262/solidity/contracts/TokenStaking.sol#L273

If the amount is non-zero, it means it is still in `TokenStaking` contract and we should include it.

Reply:
a467f405b1109378314bf29da9071bd3147d3cb2